### ISSUE TITLE
Support Array Interface (__array_interface__ attribute)

### ISSUE THREAD

**Issue by pearu** [pytorch/pytorch#54138]
## 🚀 Feature

Implement `torch.Tensor.__array_interface__` attribute as defined by [Array Interface](https://numpy.org/doc/stable/reference/arrays.interface.html)

## Motivation

The array interface allows no-copy sharing of data buffers in between different array-like objects.

PyTorch currently implements `torch.Tensor.__cuda_array_interface__` attribute defined by [CUDA Array Interface](https://numba.pydata.org/numba-doc/dev/cuda/cuda_array_interface.html) for conveniently sharing data stored in CUDA devices while sharing data in CPU memory is currently possible via NumPy ndarray object:
```
>>> import torch
>>> t = torch.tensor([1, 2, 3, 4, 5])
>>> import numpy
>>> numpy.asarray(t).__array_interface__
{'data': (94135211131264, False), 'strides': None, 'descr': [('', '<i8')], 'typestr': '<i8', 'shape': (5,), 'version': 3}
```
Dependence on NumPy for sharing data buffers between PyTorch and other array-like objects is unnecessary.

## Pitch

Add a new attribute `__array_interface__` to a `torch.Tensor` object that
- is a dictionary containing data buffer data as defined by the Array Interface.
- is enabled only when the Tensor object uses CPU storage. This behavior is in line with the `__cuda_array_interface__` attribute availability (that is enabled only when the Tensor object uses CUDA storage).
- is backward compatible to existing `torch.Tensor.__array__()` method because `numpy.asarray` prefers to use `__array_interface__` protocol over the `__array__` protocol.

## Alternatives

- The alternative is to wrap PyTorch Tensor object with NumPy ndarray as described above.
- Array Interface defines `__array_struct__` as an alternative way to share data buffers. `__array_struct__` can be considered obsolete via [PEP 3118- Buffer Protocol](https://www.python.org/dev/peps/pep-3118/).

## Additional context

This feature request is a follow-up to #51156  and https://pearu.github.io/array_interface_pytorch.html discussions.

**Comment by pearu** [comment]
As noted in https://github.com/pytorch/pytorch/issues/51156#issuecomment-791998344, exposing pytorch Tensor data is problematic if the storage of a Tensor object is changed in-place. For example:
```
>>> a = torch.tensor([1, 2, 3])
>>> x = a.__array__()
>>> x
array([1, 2, 3])
>>> a.set_(1 + a)
tensor([2, 3, 4])
>>> x
array([94375335414656, 94375291240464,              3])
>>> x.base
tensor([2, 3, 4])
```
Notice that after `a` storage is reset, the wrapped Numpy array `x` will contain garbage.

**Comment by rgommers** [comment]
Agreed, this makes sense in principle to add for design completeness. It's probably not high-prio, since the only thing it lets people do is avoid a numpy dependency. Currently they can use `__array__` or dlpack for zero-copy data sharing.

> is backward compatible to existing `torch.Tensor.__array__()` method because `numpy.asarray` prefers to use `__array_interface__` protocol over the `__array__` protocol.

This is the question. Your reasoning here is a little backward. This has the potential to break backwards compat, because `np.asarray(a_tensor)` will prefer the new `__array_interface__` attribute over `Tensor.__array__`.

It's not clear to me that what I pointed out in https://github.com/pytorch/pytorch/issues/51156#issuecomment-791998344, ```
  // TODO: This attempts to keep the underlying memory alive by setting the base
  // object of the ndarray to the tensor and disabling resizes on the storage.
  // This is not sufficient. For example, the tensor's storage may be changed
  // via Tensor.set_, which can free the underlying memory.
```
can be dealt with in `__array_interface__ in the same manner.
